### PR TITLE
Feature/azure

### PR DESCRIPTION
**Pull request by denis-yuen**
- Adds Azure support for deployer, reaper
- needs tests

### CODE REVIEW

File: youxia-common/src/main/java/io/cloudbindle/youxia/azure/resourceManagerWrapper/AzureResourceManagerClient.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+    private AuthenticationResult getAccessTokenFromUserCredentials() throws Exception {
+        // this user is an "organization user" microsoft accounts do not work
+        // see http://blog.baslijten.com/create-an-organizational-account-to-administrate-azure-when-having-a-microsoft-account/
+        // the password is set by logging in as the organization user (which will prompt you to set the password)
+        HierarchicalINIConfiguration youxiaConfig = ConfigTools.getYouxiaConfig();
+        String username = youxiaConfig.getString(ConfigTools.YOUXIA_AZURE_ACTIVE_DIRECTORY_USERNAME);
+        String password = youxiaConfig.getString(ConfigTools.YOUXIA_AZURE_ACTIVE_DIRECTORY_PASSWORD);
+        // the end part of this is from the URL of the active directory page, seriously
+        String authority = "https://login.windows.net/" + youxiaConfig.getString(ConfigTools.YOUXIA_AZURE_ACTIVE_DIRECTORY_TENANT_ID);
```
Comment (SolomonShorser-OICR) [inline youxia-common/src/main/java/io/cloudbindle/youxia/azure/resourceManagerWrapper/AzureResourceManagerClient.java]: This URL  ("login.windows.net") will  _probably_ never change... ;)

File: youxia-common/src/main/java/io/cloudbindle/youxia/azure/resourceManagerWrapper/AzureResourceManagerClient.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+    private String getEndpoint(String endpoint) throws Exception {
+        String accessToken;
+        HierarchicalINIConfiguration youxiaConfig = ConfigTools.getYouxiaConfig();
+        AuthenticationResult accessTokenFromUserCredentials = getAccessTokenFromUserCredentials();
+        accessToken = accessTokenFromUserCredentials.getAccessToken();
+
+        HttpClient client = new DefaultHttpClient();
+        HttpGet request = new HttpGet("https://management.azure.com/subscriptions/" + youxiaConfig.getString(YOUXIA_AZURE_SUBSCRIPTION_ID)
+                + "/" + endpoint + "?api-version=2015-01-01");
```
Comment (SolomonShorser-OICR) [inline youxia-common/src/main/java/io/cloudbindle/youxia/azure/resourceManagerWrapper/AzureResourceManagerClient.java]: What happens if the API version changes? Are there actually different API version avilable?

File: youxia-common/src/main/java/io/cloudbindle/youxia/azure/resourceManagerWrapper/AzureResourceManagerClient.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        // the end part of this is from the URL of the active directory page, seriously
+        String authority = "https://login.windows.net/" + youxiaConfig.getString(ConfigTools.YOUXIA_AZURE_ACTIVE_DIRECTORY_TENANT_ID);
+        // this is listed under the applications tab and is listed as CLIENT ID
+        String clientID = youxiaConfig.getString(ConfigTools.YOUXIA_AZURE_ACTIVE_DIRECTORY_CLIENT_ID);
+
+        AuthenticationContext context;
+        AuthenticationResult result = null;
+        ExecutorService service = null;
+        try {
+            service = Executors.newFixedThreadPool(1);
```
Comment (SolomonShorser-OICR) [inline youxia-common/src/main/java/io/cloudbindle/youxia/azure/resourceManagerWrapper/AzureResourceManagerClient.java:74]: Why is this running in a separate thread?

File: youxia-common/src/main/java/io/cloudbindle/youxia/azure/resourceManagerWrapper/AzureResourceManagerClient.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        String endpoint = patchEndpoint("resourcegroups/" + name, gson.toJson(patch));
+        ResourceGroup fromJson = gson.fromJson(endpoint, ResourceGroup.class);
+        return fromJson;
+    }
+
+    public static void main(String[] args) {
+        AzureResourceManagerClient client = new AzureResourceManagerClient();
+        Gson gson = new GsonBuilder().setPrettyPrinting().create();
+        try {
+            System.out.println("List of resource groups");
```
Comment (SolomonShorser-OICR) [inline youxia-common/src/main/java/io/cloudbindle/youxia/azure/resourceManagerWrapper/AzureResourceManagerClient.java]: Logging...

File: youxia-common/src/main/java/io/cloudbindle/youxia/azure/resourceManagerWrapper/AzureResourceManagerClient.java (hunk 3 above)
Comment (denis-yuen) [inline youxia-common/src/main/java/io/cloudbindle/youxia/azure/resourceManagerWrapper/AzureResourceManagerClient.java:74]: Courtesy of the microsoft, the AuthenticationContext had to be multi-threaded

**Pull request merged**